src/cargo/sources/git/oxide.rs

```diff
@@ -34,28 +34,93 @@ pub fn with_retry_and_progress(
-
-    let should_interrupt = AtomicBool::new(false);
-    let _progress_bar = Progress::new("Fetch", config);
     std::thread::scope(move |s| {
+        let progress_root: Arc<git::progress::tree::Root> = git::progress::tree::root::Options {
+            initial_capacity: 10,
+            message_buffer_capacity: 10,
+        }
+        .into();
+        let mut progress = progress_root.add_child("operation");
+        let program_aborts_on_interrupt = AtomicBool::default();
+        let mut progress_bar = Progress::new("Fetch", config);
         s.spawn({
             let root = Arc::downgrade(&progress_root);
             move || -> CargoResult<()> {
+                let read_pack_bytes: git::progress::Id =
+                    git::odb::pack::bundle::write::ProgressId::ReadPackBytes.into();
+                let delta_index_objects: git::progress::Id =
+                    git::odb::pack::index::write::ProgressId::IndexObjects.into();
+                let resolve_objects: git::progress::Id =
+                    git::odb::pack::index::write::ProgressId::ResolveObjects.into();
+
+                // We choose `N=10` here to make a `300ms * 10slots ~= 3000ms`
+                // sliding window for tracking the data transfer rate (in bytes/s).
+                let mut last_update = Instant::now();
+                let mut counter = MetricsCounter::<10>::new(0, last_update);
+
                 let mut tasks = Vec::with_capacity(10);
+                let update_interval = std::time::Duration::from_millis(300);
+                let short_check_interval = Duration::from_millis(8);
+
                 while let Some(root) = root.upgrade() {
+                    if last_update.elapsed() < update_interval {
+                        std::thread::sleep(short_check_interval);
+                        continue;
+                    }
+                    last_update = Instant::now();
                     root.sorted_snapshot(&mut tasks);
-                    // dbg!(&tasks);
-                    std::thread::sleep(Duration::from_millis(300));
+                    let counters = tasks
+                        .iter()
+                        .find_map(|(_, t)| {
+                            (t.id == read_pack_bytes).then(|| t.progress.as_ref().expect("set"))
+                        })
+                        .and_then(|read| {
+                            tasks.iter().find_map(|(_, t)| {
+                                (t.id == delta_index_objects)
+                                    .then(|| (t.progress.as_ref().expect("set"), Some(read)))
+                            })
+                        })
+                        .or_else(|| {
+                            tasks.iter().find_map(|(_, t)| {
+                                (t.id == resolve_objects)
+                                    .then(|| (t.progress.as_ref().expect("set"), None))
+                            })
+                        });
```

Comment:
When I start to see this cimplicated if a level of combinators, it makes me question the use of combinators or at least find a clearer way of expressing them.

Any thoughts on a way to simplify this?

Reply:
Thanks for pushing this!

I did give it another shot and came up with a something that reads more clearly in a9ba6b2e71bae6ae27d09161578d4c43a6d8bd4b. And there is even one more pass on top of it: 04558fb4c5d232cf11629021c9e13c0661cc8304 .

What do you think?